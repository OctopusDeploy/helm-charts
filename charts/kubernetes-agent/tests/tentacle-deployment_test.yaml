suite: "tentacle deployment"
templates:
  - templates/tentacle-deployment.yaml
tests:
  - it: "should match snapshot"
    asserts:
      - matchSnapshot: {}
  - it: "includes the watchdog when NFS and Watchdog are both enabled"
    set:
      persistence:
        storageClassName: ""
        nfs:
          watchdog:
            enabled: true
            image:
              repository: octopusdeploy/kubernetes-agent-nfs-watchdog
              tag: "0.0.1"
    asserts:
      - hasDocuments:
          count: 1
      - isKind:
          of: Deployment
      - isAPIVersion:
          of: apps/v1
      - equal:
          path: spec.template.spec.containers[0].name
          value: octopus-agent-tentacle
      - contains:
          path: spec.template.spec.containers[0].env
          content:
            name: "OCTOPUS__K8STENTACLE__NFSWATCHDOGIMAGE"
            value: "octopusdeploy/kubernetes-agent-nfs-watchdog:0.0.1"
          count: 1
      - equal:
          path: spec.template.spec.containers[1].name
          value: nfs-watchdog
      - equal:
          path: spec.template.spec.containers[1].image
          value: "octopusdeploy/kubernetes-agent-nfs-watchdog:0.0.1"

  - it: "Doesn't include the watchdog when only the Watchdog is enabled"
    set:
      persistence:
        storageClassName: "my-thing"
        nfs:
          watchdog:
            enabled: true
            image:
              repository: octopusdeploy/kubernetes-agent-nfs-watchdog
              tag: "0.0.1"
    asserts:
      - hasDocuments:
          count: 1
      - isKind:
          of: Deployment
      - isAPIVersion:
          of: apps/v1
      - equal:
          path: spec.template.spec.containers[0].name
          value: octopus-agent-tentacle
      - notContains:
          path: spec.template.spec.containers[0].env
          content:
            name: "OCTOPUS__K8STENTACLE__NFSWATCHDOGIMAGE"
      - notContains:
          path: spec.template.spec.containers
          content:
            name: "nfs-watchdog"

  - it: "Doesn't include the watchdog when watchdog is disabled"
    set:
      persistence:
        storageClassName: ""
        nfs:
          watchdog:
            enabled: false
            image:
              repository: octopusdeploy/kubernetes-agent-nfs-watchdog
              tag: "0.0.1"
    asserts:
      - hasDocuments:
          count: 1
      - isKind:
          of: Deployment
      - isAPIVersion:
          of: apps/v1
      - equal:
          path: spec.template.spec.containers[0].name
          value: octopus-agent-tentacle
      - notContains:
          path: spec.template.spec.containers[0].env
          content:
            name: "OCTOPUS__K8STENTACLE__NFSWATCHDOGIMAGE"
      - notContains:
          path: spec.template.spec.containers
          content:
            name: "nfs-watchdog"

  - it: "Sets the PERSISTENTVOLUMESIZE variable to the correct value"
    set:
      persistence:
        size: 100Gi
    asserts:
      - hasDocuments:
          count: 1
      - isKind:
          of: Deployment
      - isAPIVersion:
          of: apps/v1
      - equal:
          path: spec.template.spec.containers[0].name
          value: octopus-agent-tentacle
      - contains:
          path: spec.template.spec.containers[0].env
          content:
            name: "OCTOPUS__K8STENTACLE__PERSISTENTVOLUMESIZE"
            value: "100Gi"

  - it: "sets the tentacle pod resources from values"
    set:
      agent:
        resources:
          requests: null
          limits:
            cpu: "200m"
    asserts:
      - equal:
          path: spec.template.spec.containers[0].resources
          value:
            limits:
              cpu: "200m"
              
  - it: "truncates long names"
    set:
      nameOverride: thisisaverylongstring34087236548972364802367y408923604872341234141
    asserts:
      - equal:
          path: metadata.name
          value: thisisaverylongstring34087236548972364802367y408923604872341234-tentacle

  - it: "mounts octopus server certificate when server certificate provided"
    set:
      agent:
        serverCertificate: "dGVzdC10ZXh0"
    asserts:
      - exists:
          path: spec.template.spec.volumes[1]
      
      - equal:
          path: spec.template.spec.volumes[1]
          value:
            name: octopus-server-certificate-configmap
            configMap:
              name: "octopus-server-certificate"

      - exists:
          path: spec.template.spec.containers[0].volumeMounts[1]

      - equal:
          path: spec.template.spec.containers[0].volumeMounts[1]
          value:
            name: octopus-server-certificate-configmap
            mountPath: /etc/ssl/certs/octopus-server-certificate.pem
            subPath: octopus-server-certificate.pem
            readOnly: false
  
  - it: "adds image pulls secrets to deployment"
    set:
      imagePullSecrets:
      - name: secret-1
      - name: secret-2
    asserts:
      - equal:
          path: spec.template.spec.imagePullSecrets
          value:
            - name: secret-1
            - name: secret-2

  - it: "Changes affinity if changed"
    set:
      agent:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: kubernetes.io/arch
                  operator: In
                  values:
                  - arm64
                  - amd64
    asserts:
      - equal:
          path: spec.template.spec.affinity
          value: 
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                - matchExpressions:
                  - key: kubernetes.io/arch
                    operator: In
                    values:
                    - arm64
                    - amd64

  - it: "Sets pod tolerations if defined"
    set:
      agent:
        tolerations:
        - key: "tol-1"
          operator: "Equal"
          value: "val-1"
          effect: "NoSchedule"
        - key: "tol-21"
          operator: "Equal"
          value: "val-2"
          effect: "NoSchedule"
    asserts:
      - equal:
          path: spec.template.spec.tolerations
          value:
          - key: "tol-1"
            operator: "Equal"
            value: "val-1"
            effect: "NoSchedule"
          - key: "tol-21"
            operator: "Equal"
            value: "val-2"
            effect: "NoSchedule"

  - it: "Sets pod security context if defined"
    set:
      agent:
        securityContext:
          runAsUser: 10101
          runAsGroup: 234
          fsGroup: 545
    asserts:
      - equal:
          path: spec.template.spec.securityContext
          value:
            runAsUser: 10101
            runAsGroup: 234
            fsGroup: 545
  
  - it: "includes the tag suffix in the agent image tag when defined"
    set:
      agent:
        image:
          repository: "myImageRepository"
          pullPolicy: IfNotPresent
          tag: "1.2.3"
          tagSuffix: "bullseye-slim"
    asserts:
      - equal:
          path: spec.template.spec.containers[0].image
          value: "myImageRepository:1.2.3-bullseye-slim"
          
  - it: "uses default agent image with suffix when tag is undefined"
    set:
      agent:
        image:
          pullPolicy: IfNotPresent
          tag: ""
          tagSuffix: "bullseye-slim"
    asserts:
      - equal:
          path: spec.template.spec.containers[0].image
          value: "octopusdeploy/kubernetes-agent-tentacle:8.2.2504-bullseye-slim"
