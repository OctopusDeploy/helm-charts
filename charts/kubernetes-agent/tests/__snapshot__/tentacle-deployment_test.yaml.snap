should match snapshot:
  1: |
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      labels:
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: octopus-agent
        app.kubernetes.io/version: 8.2.2504
        helm.sh/chart: kubernetes-agent-1.18.2
      name: octopus-agent-tentacle
      namespace: NAMESPACE
    spec:
      replicas: 1
      selector:
        matchLabels:
          app.kubernetes.io/instance: RELEASE-NAME
          app.kubernetes.io/name: octopus-agent
      template:
        metadata:
          labels:
            app.kubernetes.io/instance: RELEASE-NAME
            app.kubernetes.io/managed-by: Helm
            app.kubernetes.io/name: octopus-agent
            app.kubernetes.io/version: 8.2.2504
            helm.sh/chart: kubernetes-agent-1.18.2
        spec:
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                  - matchExpressions:
                      - key: kubernetes.io/os
                        operator: In
                        values:
                          - linux
                      - key: kubernetes.io/arch
                        operator: In
                        values:
                          - arm64
                          - amd64
          containers:
            - env:
                - name: ACCEPT_EULA
                  value: "N"
                - name: AgentName
                  value: ""
                - name: DeploymentTargetEnabled
                  value: "true"
                - name: ServerCommsAddresses
                  value: ""
                - name: ServerCommsAddress
                  value: ""
                - name: ServerUrl
                  value: ""
                - name: Space
                  value: Default
                - name: TargetEnvironment
                  value: ""
                - name: TargetRole
                  value: ""
                - name: TargetTenant
                  value: ""
                - name: TargetTenantTag
                  value: ""
                - name: TargetTenantedDeploymentParticipation
                  value: Untenanted
                - name: DefaultNamespace
                  value: ""
                - name: OCTOPUS__K8STENTACLE__NAMESPACE
                  value: NAMESPACE
                - name: OCTOPUS__K8STENTACLE__PODSERVICEACCOUNTNAME
                  value: octopus-agent-scripts
                - name: OCTOPUS__K8STENTACLE__PODVOLUMECLAIMNAME
                  value: octopus-agent-nfs-release-name-pvc-10gi
                - name: OCTOPUS__K8STENTACLE__HELMRELEASENAME
                  value: RELEASE-NAME
                - name: OCTOPUS__K8STENTACLE__HELMCHARTVERSION
                  value: 1.18.2
                - name: OCTOPUS__K8STENTACLE__DISABLEAUTOPODCLEANUP
                  value: "false"
                - name: OCTOPUS__K8STENTACLE__DISABLEPODEVENTSINTASKLOG
                  value: "false"
                - name: OCTOPUS__K8STENTACLE__NFSWATCHDOGIMAGE
                  value: octopusdeploy/kubernetes-agent-nfs-watchdog:0.2.0
                - name: OCTOPUS__TENTACLE__LOGLEVEL
                  value: Info
                - name: TentacleHome
                  value: /octopus
                - name: TentacleApplications
                  value: /octopus/Applications
                - name: TentaclePollingConnectionCount
                  value: "5"
                - name: OCTOPUS__K8STENTACLE__ENABLEMETRICSCAPTURE
                  value: "true"
                - name: OCTOPUS__K8STENTACLE__PODRESOURCEJSON
                  value: '{"requests":{"cpu":"25m","memory":"100Mi"}}'
                - name: OCTOPUS__K8STENTACLE__PODAFFINITYJSON
                  value: '{"nodeAffinity":{"requiredDuringSchedulingIgnoredDuringExecution":{"nodeSelectorTerms":[{"matchExpressions":[{"key":"kubernetes.io/os","operator":"In","values":["linux"]},{"key":"kubernetes.io/arch","operator":"In","values":["arm64","amd64"]}]}]}}}'
                - name: OCTOPUS__K8STENTACLE__PERSISTENTVOLUMESIZE
                  value: 10Gi
              image: octopusdeploy/kubernetes-agent-tentacle:8.2.2504
              imagePullPolicy: IfNotPresent
              name: octopus-agent-tentacle
              resources:
                requests:
                  cpu: 100m
                  memory: 150Mi
              startupProbe:
                exec:
                  command:
                    - cat
                    - /etc/octopus/initialized
                failureThreshold: 100
                initialDelaySeconds: 2
                periodSeconds: 2
              volumeMounts:
                - mountPath: /octopus
                  name: tentacle-home
            - env:
                - name: watchdog_directory
                  value: /octopus
              image: octopusdeploy/kubernetes-agent-nfs-watchdog:0.2.0
              imagePullPolicy: IfNotPresent
              name: nfs-watchdog
              volumeMounts:
                - mountPath: /octopus
                  name: tentacle-home
          serviceAccountName: octopus-agent-tentacle
          volumes:
            - name: tentacle-home
              persistentVolumeClaim:
                claimName: octopus-agent-nfs-release-name-pvc-10gi
